main.go

```diff
@@ -0,0 +1,388 @@
+	default:
+		return fmt.Errorf("failed to interpret base as index or image: %v", baseDesc.MediaType)
+	}
+	baseIndex, err := baseDesc.ImageIndex()
+	if err != nil {
+		return err
+	}
+
+	im, err := baseIndex.IndexManifest()
+	if err != nil {
+		return fmt.Errorf("failed to interpret base as index: %w", err)
+	}
+	var adds []mutate.IndexAddendum
+	// Try to build images for all supported platforms.
+	for _, id := range im.Manifests {
+		logf := withPrefix(logf, fmt.Sprintf("%v/%v: ", id.Platform.OS, id.Platform.Architecture))
+		if err := verifyPlatform(*id.Platform); err != nil {
+			logf("skipping: %v", err)
+			continue
+		}
+		logf("base digest: %v", id.Digest)
+		bi, err := baseIndex.Image(id.Digest)
+		if err != nil {
+			return err
+		}
+		logf("building")
+		img, err := createImageForBase(logf, staticFiles, gopaths, ldFlags, bi)
+		if err != nil {
+			return err
+		}
+		d, err := img.Digest()
+		if err != nil {
+			return err
+		}
+		logf("new digest: %v", d)
+		adds = append(adds, mutate.IndexAddendum{
+			Add: img,
+			Descriptor: v1.Descriptor{
+				MediaType:   id.MediaType,
+				URLs:        id.URLs,
+				Annotations: id.Annotations,
+				Platform:    id.Platform,
+			},
+		})
+	}
+
+	// Generate a new index with all the platform images.
+	idx := mutate.AppendManifests(mutate.IndexMediaType(empty.Index, types.OCIImageIndex), adds...)
+	d, err := idx.Digest()
+	if err != nil {
+		return err
+	}
+	logf("index digest: %v", d)
+	if !push {
+		logf("not pushing")
+		return nil
+	}
+
+	for _, r := range refs {
+		logf("pushing to %v", r)
+		if err := remote.WriteIndex(r, idx, remoteOpts...); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func createImageForBase(logf logf, staticFiles, goPaths map[string]string, ldFlags string, base v1.Image) (v1.Image, error) {
+	cf, err := base.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+	tmpDir, err := os.MkdirTemp("", "mkctr")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	env := append(os.Environ(),
+		"CGO_ENABLED=0",
+		"GOOS="+cf.OS,
+		"GOARCH="+cf.Architecture,
+	)
+	if cf.Architecture == "arm" {
+		env = append(env, "GOARM=5")
+	}
+
+	files := map[string]string{}
+	for src, dst := range staticFiles {
+		files[src] = dst
+	}
+
+	// Compile all the goPaths
+	for gp, dst := range goPaths {
+		logf("compiling %v", gp)
+		n, err := compileGoBinary(gp, tmpDir, env, ldFlags)
+		if err != nil {
+			return nil, err
+		}
+		logf("output %v -> %v", gp, n)
+		files[n] = dst
+	}
+	layer, err := layerFromFiles(logf, files)
+	if err != nil {
+		return nil, err
+	}
+	return mutate.AppendLayers(base, layer)
+}
+
+func compileGoBinary(what, where string, env []string, ldFlags string) (string, error) {
+	f, err := os.CreateTemp(where, "out")
+	if err != nil {
+		return "", err
+	}
+	out := f.Name()
+	if err := f.Close(); err != nil {
+		return "", err
+	}
+	cmd := exec.Command(
+		"go",
+		"build",
+		"-v",
+		"-tags=xversion",
```

Comment:
what's xversion for?

Reply:
i copied it over from the dockerfile, not required. removed